wait fork trigger temporary split across generated C++ functions, output does not compile

### Can you please attach an example that shows the issue?

`min_wait_fork.sv`, self-contained, CC0:

```systemverilog
// SPDX-FileCopyrightText: 2026 Wilson Snyder
// SPDX-License-Identifier: CC0-1.0
module tb;
  logic clk = 1'b0;
  always #5 clk = ~clk;
  initial begin
    fork @(posedge clk); join_none
    wait fork;
    $finish;
  end
endmodule
```

### What output from that test indicates it is wrong, and what is the correct or expected output?

Expected: it compiles and runs to `$finish` at 5ps. Actual: the generated C++ does not compile.

```
Vtb___024root__2.cpp: In function 'VlCoroutine Vtb___024root___eval_initial__TOP__Vtiming__0__2(...)':
Vtb___024root__2.cpp:23:13: error: '__Vtrigprevexpr_h5d9da2ce__0' was not declared in this scope
   23 |             __Vtrigprevexpr_h5d9da2ce__0 = vlProcess->completedFork();
make: *** [verilated.mk:293: Vtb___024root__2.o] Error 1
```

The `wait fork` dynamic-trigger temporary is emitted as a function-local of one sub-function
(`...__Vtiming__0__0`) while the coroutine that assigns it is a different sub-function
(`...__Vtiming__0__2`), in a different translation unit.

Two further symptoms on git master:

- `verilator_bin_dbg --debug` reports a broken link, because `V3InlineCFuncs` inlines the
  sub-function that received the declaration, renames the local it now owns and frees the original
  `AstVar`, while the other sub-function still holds `AstVarRef`s to it:

  ```
  %Error: Internal Error: min_wait_fork.sv:6:5: ../V3Broken.cpp:182: Broken link in node
          (or something without maybePointedTo): 'm_varp && !m_varp->brokeExists()'
  ```

- Built `-O3`, `verilator_bin` **segfaults** on this input rather than emitting the bad C++.

An ordinary edge trigger is unaffected: `__Vtrigprevexpr___TOP__tb__DOT__clk__0` is a root class
member and survives any split.

### What 'verilator' command line do we use to run your example?

```sh
verilator --binary --timing --output-split-cfuncs 20 -top tb -o sim.exe min_wait_fork.sv
```

`--output-split-cfuncs 20` is only a size proxy, to force a split on a small example. It also
reproduces on **default flags** for any single process exceeding the 20000 default before its
`wait fork`, because `localizeVars()` places the declaration at the very front of the process. This
generates such a case, which fails with no split flag at all:

```python
lines = ["module tb;", "  logic clk = 1'b0;", "  int acc = 0;", "  int mem [0:4095];",
         "  always #5 clk = ~clk;", "  initial begin"]
for i in range(4096):
    lines.append(f"    mem[{i}] = {i} * 3 + acc;")
    lines.append(f"    acc = acc + mem[{i}];")
lines += ["    fork @(posedge clk); join_none", "    wait fork;",
          "    if (acc == 0) $stop;",
          "    $write(\"*-* All Finished *-*\\n\");", "    $finish;", "  end", "endmodule"]
open("big_wait_fork.sv", "w").write("\n".join(lines) + "\n")
```

```
$ verilator --binary --timing -top tb -o sim.exe big_wait_fork.sv
Vtb___024root__2.cpp:4982:13: error: '__Vtrigprevexpr_h5d9da2ce__0' was not declared in this scope
```

It was originally hit on a 64 KB generated testbench with 4 `wait fork` sites on stock flags,
producing 17 instances of the error across `Vtb_...__12.cpp` through `Vtb_...__31.cpp`.

### What 'verilator --version' are you using? Did you try it with the git master version? Did you try it with other simulators?

Reproduced on git master, `5.051 devel rev v5.050-99-gf8fb1d6` (f8fb1d6, 2026-07-24), and on
released 5.048 and 5.049-devel.

Present in every release since **5.046**. `git diff v5.050 HEAD` over `V3SchedUtil.cpp`,
`V3Timing.cpp`, `V3SenExprBuilder.h` and `V3Sched.cpp` shows only an unrelated refactor of
`hasDisableQueuePushSelfPrefix`, so 5.050 has the same defect in the same form.

Not tried on other simulators; the input is legal IEEE 1800 and Verilator accepts it through
elaboration, so this is purely a code-generation defect.

### What OS and distribution are you using?

Ubuntu 24.04 on WSL2 (Linux 6.6.87.2-microsoft-standard-WSL2), g++ 15.2.0.

### May we assist you in trying to fix this in Verilator yourself?

I have a fix and will open a PR referencing this issue.

Root cause: `V3Sched::util::splitCheck()` cuts a function's top level statement list on node count
alone, ignoring `AstVar` declarations sitting in that list. Its own
`UASSERT_OBJ(!ofuncp->varsp(), ..., "Can't split function with local variables")` shows the intent,
but only checks `AstCFunc::varsp()`, not declarations among the statements, and `localizeVars()`
puts them in `stmtsp()`.

This is the same failure as #5987, which #5822 fixed by removing function locals from
`SenExprBuilder` ("function locals are not safe here because we might need to split up the generated
function"). #6926 reintroduced them for the dynamic-trigger path, because #6859 needs the temporary
to be per-process. The two requirements conflict, so rather than removing the locals again the fix
makes `splitCheck()` honour them: a sub-function boundary is only allowed where it does not separate
a local declaration from a reference to it.

--- a/docs/CONTRIBUTORS
+++ b/docs/CONTRIBUTORS
@@ -201,6 +201,7 @@ Maarten De Braekeleer
 Maciej Sobkowski
 Marcel Chang
 Marco Bartoli
+Marco Brambilla
 Marco Widmer
 Mariusz Glebocki
 Markus Krause

--- a/src/V3SchedUtil.cpp
+++ b/src/V3SchedUtil.cpp
@@ -157,12 +157,49 @@ void splitCheckFinishSubFunc(AstCFunc* ofuncp, AstCFunc* subFuncp,
     }
 }
 
+// Compute, for each top level statement of 'ofuncp', whether a new sub-function may begin there.
+// Sub-functions are emitted as separate C++ functions, so they cannot see each other's automatic
+// storage. A function-local AstVar declared among the top level statements must therefore end up
+// in the same sub-function as every reference to it, otherwise the emitted C++ refers to an
+// undeclared identifier (and the AstVar can be deleted from under the still-live references).
+static std::vector<bool> splitCheckBreakable(const AstCFunc* ofuncp) {
+    // Gather the top level statements, and where each locally declared variable is declared
+    std::vector<const AstNode*> stmtps;
+    std::unordered_map<const AstVar*, size_t> declIdx;  // Local var -> index it is declared at
+    for (const AstNode* nodep = ofuncp->stmtsp(); nodep; nodep = nodep->nextp()) {
+        if (const AstVar* const varp = VN_CAST(nodep, Var)) declIdx.emplace(varp, stmtps.size());
+        stmtps.push_back(nodep);
+    }
+
+    // Find the last statement referencing each locally declared variable
+    std::vector<size_t> lastUse(stmtps.size());
+    for (size_t i = 0; i < stmtps.size(); ++i) {
+        lastUse[i] = i;  // A declaration is live at least where it is declared
+        stmtps[i]->foreach([&](const AstNodeVarRef* refp) {
+            const auto it = declIdx.find(refp->varp());  // 'end()' if not one of our locals
+            if (it != declIdx.end()) lastUse[it->second] = std::max(lastUse[it->second], i);
+        });
+    }
+
+    // A break before statement 'i' is allowed only if no local declared before 'i' is still live
+    std::vector<bool> breakable(stmtps.size(), true);
+    size_t liveEnd = 0;  // Last index any so far declared local is referenced at
+    for (size_t i = 0; i < stmtps.size(); ++i) {
+        breakable[i] = liveEnd < i;
+        if (VN_IS(stmtps[i], Var)) liveEnd = std::max(liveEnd, lastUse[i]);
+    }
+    return breakable;
+}
+
 // Split large function according to --output-split-cfuncs
 void splitCheck(AstCFunc* const ofuncp) {
     if (!ofuncp) return;
     UASSERT_OBJ(!ofuncp->varsp(), ofuncp, "Can't split function with local variables");
     if (!v3Global.opt.outputSplitCFuncs() || !ofuncp->stmtsp()) return;
     if (ofuncp->nodeCount() < v3Global.opt.outputSplitCFuncs()) return;
+
+    // Statement boundaries that would separate a local declaration from a reference to it
+    const std::vector<bool> breakable = splitCheckBreakable(ofuncp);
 
     // Need to find the AstVarScopes for the function arguments. They should be in the same Scope.
     std::unordered_map<const AstVar*, AstVarScope*> argVscps;
@@ -187,13 +224,13 @@ void splitCheck(AstCFunc* const ofuncp) {
 
     // Move statements one by one to the new sub-functions
     AstNode* stmtsp = ofuncp->stmtsp()->unlinkFrBackWithNext();
-    while (AstNode* const itemp = stmtsp) {
+    for (size_t index = 0; AstNode* const itemp = stmtsp; ++index) {
         stmtsp = stmtsp->nextp();
         if (stmtsp) stmtsp->unlinkFrBackWithNext();
         const size_t itemSize = static_cast<size_t>(itemp->nodeCount());
         size += itemSize;
 
-        if (size > static_cast<size_t>(v3Global.opt.outputSplitCFuncs())) {
+        if (size > static_cast<size_t>(v3Global.opt.outputSplitCFuncs()) && breakable[index]) {
             if (subFuncp) splitCheckFinishSubFunc(ofuncp, subFuncp, argVscps);
             subFuncp = nullptr;
             size = itemSize;

--- a/test_regress/t/t_timing_wait_fork_split.py
+++ b/test_regress/t/t_timing_wait_fork_split.py
@@ -1,0 +1,26 @@
+#!/usr/bin/env python3
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of either the GNU Lesser General Public License Version 3
+# or the Perl Artistic License Version 2.0.
+# SPDX-FileCopyrightText: 2026 Wilson Snyder
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+import vltest_bootstrap
+
+test.scenarios('simulator')
+
+# --output-split-cfuncs forces the process to be split into sub-functions
+test.compile(verilator_flags2=["--binary", "--output-split-cfuncs", "20"])
+
+# Confirm the split actually happened: the 'initial' timing process is emitted
+# as function sub-parts (__Vtiming__0__0 / __Vtiming__0__1). Without the split
+# this test would not exercise the bug it guards against -- a dynamic 'wait
+# fork' trigger temporary separated from its uses across a sub-function boundary.
+if test.vlt_all:
+    test.file_grep(test.obj_dir + "/V" + test.name + "_t__0.cpp", r'__Vtiming__0__1\b')
+
+test.execute()
+
+test.passes()

--- a/test_regress/t/t_timing_wait_fork_split.v
+++ b/test_regress/t/t_timing_wait_fork_split.v
@@ -1,0 +1,40 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// A 'wait fork' needs a dynamic trigger temporary, which is function-local to
+// the process it appears in. Splitting the process into sub-functions must not
+// separate that declaration from its uses, as sub-functions are emitted as
+// separate C++ functions.
+//
+// This file ONLY is placed under the Creative Commons Public Domain.
+// SPDX-FileCopyrightText: 2026 Wilson Snyder
+// SPDX-License-Identifier: CC0-1.0
+
+module t;
+  logic clk = 1'b0;
+  int   cnt = 0;
+
+  always #5 clk = ~clk;
+
+  task automatic phase(int n);
+    fork begin @(posedge clk); cnt += n; end join_none
+    wait fork;
+  endtask
+
+  initial begin
+    fork @(posedge clk); join_none
+    wait fork;
+    phase(1);
+    fork @(negedge clk); join_none
+    wait fork;
+    phase(2);
+    fork @(posedge clk); join_none
+    wait fork;
+    phase(4);
+    if (cnt != 7) begin
+      $write("%%Error: cnt=%0d exp=7\n", cnt);
+      $stop;
+    end
+    $write("*-* All Finished *-*\n");
+    $finish;
+  end
+endmodule

